packages/udt/src/udt/index.ts

```diff
@@ -1,306 +1,394 @@
-   *
-   * @returns A promise resolving to the token's name.
-   */
-  async name(
-    context?: ssri.ContextScript,
-  ): Promise<ssri.ExecutorResponse<string | undefined>> {
-    if (this.executor) {
-      const res = await this.executor.runScriptTry(this.code, "UDT.name", [], {
-        script: this.script,
-        ...context,
-      });
-      if (res) {
-        return res.map((res) => ccc.bytesTo(res, "utf8"));
-      }
-    }
+export class UdtManager {
+  constructor(
+    public readonly script: ccc.Script,
+    public readonly cellDeps: ccc.CellDep[],
+    public readonly dataLenRange: [number, number],
+    public readonly name: string,
+    public readonly symbol: string,
+    public readonly decimals: number,
+    public readonly icon: string,
+  ) {}
 
-    return ssri.ExecutorResponse.new(undefined);
-  }
+  async *find(
+    client: ccc.Client,
+    locks: ccc.Script[],
+    options?: {
+      source?: "local" | "chain";
+    },
+  ): AsyncGenerator<ccc.Cell> {
+    const isOnChain = options?.source === "chain";
+    const processedLocks = new Set<ccc.Hex>();
+    for (const lock of locks) {
+      const lockHash = lock.hash();
+      if (processedLocks.has(lockHash)) {
+        continue;
+      }
+      processedLocks.add(lockHash);
 
-  /**
-   * Retrieves the symbol of the UDT.
-   * @returns The symbol of the UDT.
-   */
-  async symbol(
-    context?: ssri.ContextScript,
-  ): Promise<ssri.ExecutorResponse<string | undefined>> {
-    if (this.executor) {
-      const res = await this.executor.runScriptTry(
-        this.code,
-        "UDT.symbol",
-        [],
+      const findCellsArgs = [
         {
-          script: this.script,
-          ...context,
+          script: lock,
+          scriptType: "lock",
+          filter: {
+            script: this.script,
+            outputDataLenRange: this.dataLenRange,
+          },
+          scriptSearchMode: "exact",
+          withData: true,
         },
-      );
-      if (res) {
-        return res.map((res) => ccc.bytesTo(res, "utf8"));
+      ] as const;
+
+      for await (const cell of isOnChain
+        ? client.findCellsOnChain(...findCellsArgs)
+        : client.findCells(...findCellsArgs)) {
+        yield cell;
       }
     }
-
-    return ssri.ExecutorResponse.new(undefined);
   }
 
-  /**
-   * Retrieves the decimals of the UDT.
-   * @returns The decimals of the UDT.
-   */
-  async decimals(
-    context?: ssri.ContextScript,
-  ): Promise<ssri.ExecutorResponse<ccc.Num | undefined>> {
-    if (this.executor) {
-      const res = await this.executor.runScriptTry(
-        this.code,
-        "UDT.decimals",
-        [],
-        {
-          script: this.script,
-          ...context,
-        },
-      );
-      if (res) {
-        return res.map((res) => ccc.numFromBytes(res));
+  // eslint-disable-next-line @typescript-eslint/require-await
+  async infoFrom(
+    _client: ccc.Client,
+    initialInfo = {
```

Comment:
I personally dislike this design. My principle is that we should always obtain information from the original `Transaction` (the original one defined by CKB, not CCC's) if possible. An example is the `spore` package, where all cobuild things are hidden behind the `Transaction.witnesses`.

Yes, extra information does make implementation simpler, and it will likely improve performance. However, the increased complexity is too painful. It's a trade-off between performance/simplicity, and usability in most cases.

Reply:
@Hanssen0 where in the original PR was mentioned that CoBuild is a target?

>  we should always obtain information from the original Transaction

Let's break down what it entails:
- **inputs**: the client is exactly there as initial parameter for obtaining information about the original Transaction, like the inclusion block and **Header**.
- **outputs**: we need to update the parameter signature.

```typescript
    ...cells: {
      cellOutput: ccc.CellOutput;
      outputData: ccc.Hex;
      outPoint?: ccc.OutPoint;
    }[]
```

becomes:

```typescript
    ...cells: ccc.Cell[] | {tx : ccc.Transaction, index: number}[]
```

Not really an issue as you can see.

Reply:
Let me clarify: the `Info` is an extra thing that devs must maintain. When using the API, invokers must keep the info along with the tx, even if we can get all we need from the tx. It makes things complex.

Reply:
Underlying issue is that CCC doesn't offer a good way to attach additional information to entities like Transaction, Cell and Script... As you know I have been struggling with this since the beginning due to the iCKB headers.

I recently switched to client+cache for the additional information and I was hoping that it would be the end of it.

The CoBuild info would be the object that later on gets serialized into the message?

We can just use a global CCC WeakMap to add this extra info, you just need to modify clone and copy to persist this information across copies

Reply:
Besides letting devs focus their mind on a simple tx, this also gives extra compatibility across SDK/ABI/API. Imagine that we're dealing with two devices cooperating to build a transaction. If we have additional information attached to the transaction, we will need to design the ABI carefully to carry it. Alternatively, we need to "pack" this information into the transaction and then "unpack" it on the other end.

Both ways are annoying & complex & leaking abstraction. That's why I insist that it's necessary to avoid attaching information to the transaction.